storage/mysql/log_storage.go

```diff
@@ -500,6 +513,52 @@ func (t *logTreeTX) QueueLeaves(ctx context.Context, leaves []*trillian.LogLeaf,
 	return existingLeaves, nil
 }
 
+func (t *logTreeTX) AddSequencedLeaves(ctx context.Context, leaves []*trillian.LogLeaf) ([]*trillian.QueuedLogLeaf, error) {
+	res := make([]*trillian.QueuedLogLeaf, len(leaves))
+	ok := status.New(codes.OK, "OK").Proto()
+
+	// Note: Leaves are sorted by LeafIndex, so no deterministic reordering is
+	// necessary to avoid deadlocks.
+	for i, leaf := range leaves {
+		// This should fail on insert, but catch it early.
+		if got, want := len(leaf.LeafIdentityHash), t.hashSizeBytes; got != want {
+			return nil, status.Errorf(codes.FailedPrecondition, "leaves[%d] has incorrect hash size %d, want %d", i, got, want)
+		}
+
+		res[i] = &trillian.QueuedLogLeaf{Status: ok}
+
+		// TODO(pavelkalinnikov): Measure latencies.
+		_, err := t.tx.ExecContext(ctx, insertLeafDataSQL,
+			t.treeID, leaf.LeafIdentityHash, leaf.LeafValue, leaf.ExtraData, 0)
+		// Note: QueueTimestamp == 0 because the entry bypasses the queue.
+		// TODO(pavelkalinnikov): Fix integration latency metrics.
+
+		// TODO(pavelkalinnikov): Support opting out from duplicates detection.
+		if isDuplicateErr(err) {
+			res[i].Status = status.New(codes.FailedPrecondition, "conflicting LeafIdentityHash").Proto()
+			continue
+		} else if err != nil {
+			glog.Warningf("Error inserting leaves[%d] into LeafData: %s", i, err)
+			return nil, err
+		}
+
+		_, err = t.tx.ExecContext(ctx, insertSequencedLeafSQL,
+			t.treeID, leaf.LeafIdentityHash, leaf.MerkleLeafHash, leaf.LeafIndex, 0)
+		// TODO(pavelkalinnikov): Update IntegrateTimestamp on integrating the leaf.
+
+		if isDuplicateErr(err) {
```

Comment:
Started manually testing, found a bug. Suppose we are trying to erroneously insert a new unique identity to an occupied leaf index. The `INSERT` above stores the identity, but the second `INSERT` fails because the index is occupied. Still, the `tx` gets committed, and the side-effect remains. Now this identity can't be inserted anymore.

I think I should add a test for this, and couple of simpler scenarios.

Reply:
Wow, this actually complicates things more than I thought it would. Essentially, we need to do 2 `INSERT`s, each can fail independently of the other (it can be conflicting identity and/or conflicting leaf index). Thus, if one fails, we should cancel the effect of the other to leave the db in the old state. We can do it by either making separate transactions, which is slow, or by manually deleting inserted keys which doesn't look nice.

@AlCutter @Martin2112 Any thoughts?

Reply:
The third option is reading before doing the inserts.

Reply:
Let's look at this again next week. Might be able to use savepoints for this at least in MySQL.

Reply:
Yes, this works. Will update the PR shortly.